### PR TITLE
Optionally floodfill empty space, allow choice of inside of mesh being free or occupied

### LINKED ISSUE #3: Maps generated using this repo not playing well with mav_voxblox_planning
Hi, I am trying to generate TSDF maps from Gazebo worlds using this repo (either with the provided building.world or my own custom worlds), which I am then trying to use with the voxblox_rrt_planner repo. While the planner works with its own sample datasets such as shed, rubble etc.; the maps generated using this repo do not work with the planner.

Essentially, the planner complains that no start/goal positions are valid for the robot, although they lie in free space. Most of the time, voxblox reports the output of getMapDistance() for these points as zero. Additionally, when compared to the planner's sample datasets, the maps generated from this repo do not have a significant "traversable" pointcloud. Sample picture here, where the ESDF pointcloud can be seen (green points), along with the "traversability" point cloud (bigger rainbow squares, you can see there are very few). The start and goal locations shown here are considered to be "occupied" by voxblox. 

image

My workflow is as follows:

1. Run voxblox_ground_truth, save TSDF.
2. Run voxblox_rrt_planner and load TSDF, rely on the internal conversion to ESDF.
3. Attempt to use the planner service with the newly generated ESDF. 

Am I missing a step when generating these maps? As voxblox_rrt_planner works with the sample datasets provided in its own repo, I decided to post the issue here.

### PR DESCRIPTION
**Pull request by helenol**
# Changes
 * New parameter for SDF builder: `fill_inside`, which determines if (for a *watertight mesh*) the inside is supposed to be free or occupied. Examples:
   * Stanford bunny: inside is occupied, since it's a single object that you're looking at from *outside*.
   * Scan of an office: inside is NOT occupied, since the mesh represents the walls of the entire office and you're looking from *inside*.
 * New option to fill any free space in the TSDF with a fixed value. This basically performs a flood-fill within the axis aligned bounding box (AABB) of the entire allocated TSDF, setting all unallocated/unobserved voxels with *any* free neighbors to a fixed distance, 4*voxel size by default.
 * Make slice height in the visualization a ROS param
 

Should hopefully fix https://github.com/ethz-asl/voxblox_ground_truth/issues/3

# Examples on Office Mesh

Watertight mesh of 3D reconstruction of an office (MeshLab with xray shader so you can actually see the office-y bits inside):
image

Behavior before (`fill_inside = true` `floodfill_unoccupied = false`):
image

With `fill_inside = false` and `floodfill_unoccupied = false`
image

With `fill_inside = false` and `floodfill_unoccupied = true`
image

Stanford bunny with floodfill (note the error on the edge... it's a result of how we iterate over the neighbors in a lazy way rather than proper floodfill :/ ):

image

Let me know if this feature is useful then I'll fix the floodfilling edge.

### CODE REVIEW

File: launch/ply_importer_demo.launch [hunk 1]
```diff
@@ -8,8 +8,12 @@
   <arg name="rotation" default="0 0 0.7071068 0.7071068" />
 
   <node name="ply_to_sdf" pkg="voxblox_ground_truth" type="ply_to_sdf" output="screen"
-        args="$(arg ply_input_file_path) $(arg tsdf_output_file_path) $(arg voxel_size) $(arg scale) $(arg translation) $(arg rotation)"/>
+        args="$(arg ply_input_file_path) $(arg tsdf_output_file_path) $(arg voxel_size) $(arg scale) $(arg translation) $(arg rotation)">
+    <param name="fill_inside" value="true" />
+    <param name="floodfill_unoccupied" value="false" />
+    <param name="tsdf_slice_height" value="4" />
```
Comment (victorreijgwart) [inline launch/ply_importer_demo.launch]: It'd nice if these settings were also available through the command-line. I can take care of this.
@helenol what is your opinion on using GFlags? That could be a good alternative to the [current simple implementation](https://github.com/ethz-asl/voxblox_ground_truth/blob/master/src/user_interfaces/ply_to_sdf_script.cpp#L25) which assumes a fixed order etc.

**Review by victorreijgwart — COMMENTED** [review]
Looks great Helen! Thank you for implementing this.

Regarding the flood fill, is the current problem due to the order in which you iterate over the voxels (i.e. with a triple for loop instead of propagating a wavefront from a seed point)?

File: launch/ply_importer_demo.launch (hunk 1 above)
Comment (helenol) [reply launch/ply_importer_demo.launch]: Yeah for command-line tools for sure. Just depends if you want this to be a ROS util or not (if you commit to ROS, i.e., for visualization anyway, then you can always do _fill_inside:=true from the command-line for example). If a non-ROS tool, gflags are great but they're a pain to use from inside ROS.

**Comment by helenol** [comment]
> Looks great Helen! Thank you for implementing this.
> 
> Regarding the flood fill, is the current problem due to the order in which you iterate over the voxels (i.e. with a triple for loop instead of propagating a wavefront from a seed point)?

Yes absolutely. Again the "lazy" way of doing it. ;) But actually can be easily fixed since you actually know that if there are no neighbors = you're outside the mesh structure, so `fill_inside` will tell you the value. I'll implement that before I merge.

FYI I also updated the CMakeLists to expose the targets properly for linking (using this as a library within antoher thing...)

**Comment by helenol** [comment]
Here's the new/fixed version, note the lack of cut-off corner:
image

**Pull request merged**